### ISSUE TITLE
Finetune Harris WGAN for the wind downscaling task

### ISSUE THREAD

**Issue by mlangguth89** [mlangguth89/downscaling_benchmark#44]
Issue-branch to document fine-tuning of the Harris WGAN model for the wind downscaling task.
For the temperature downscaling task, the following [hedgedoc](https://gitlab.jsc.fz-juelich.de/hedgedoc/SPoqQb4ZRzOP3XUyA4-cBA?edit) is available where experiment 7  constitues the configuration used for the 2m temperature downscaling benchmark model.

**Comment by seblehner** [comment]
Starting experiments as suggested by @mlangguth89 :
- LRs as in `hedgedoc #8` $l_{r,gen} = 10e-5$, $l_{r,cri} = 10e-7$ (same config as t2m downscaling), `recon_weight` starting at 1000, but test increase to 1500
- replace final layer with softplus layer and min-max norm as with the original Harris WGAN implementation => **only if necessary and other experiments fail**

Experiments overview:

| Exp.No. | Hyperparam specification | Status | Notes |
|---|---|---|---|
| 1  | $l_{r,gen} = 10^{-5}$ <br> $l_{r,cri} = 5^{-7}$ <br> $recon_{weight}=1000$  | finished 4 epochs |  optimal t2m config ($jobid=10778296$) |
| 2 | same as 1, but with <br> $recon_{weight}=1500$  | finished 4 epochs  | ($jobid=10778349$) |
|   |   |   |  |

**Comment by seblehner** [comment]
### Experiment 1

jobid = 10778296

| Epoch | c_loss | gp_loss | d_loss | cg_loss | recon_loss | g_loss | val_cg_loss | val_recon_loss |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Epoch 1 |  -2.8798 |  9.6199 |  6.7401 |  92.4105 |  278.6011 |  371.0116 |  134.5204 |  222.7058 |
| Epoch 2 |  -3.6031 |  9.2796 |  5.6765 |  286.4589 |  255.5963 |  542.0552 |  448.5300 |  257.9943 |
| Epoch 3 |  -3.9306 |  8.4589 |  4.5283 |  490.0517 |  245.3462 |  735.3979 |  363.3390 |  210.5852 |
| Epoch 4 |  -3.2354 |  7.2021 |  3.9667 |  608.2070 |  242.4057 |  850.6127 |  574.1439 |  217.9296 |

### Experiment 2

jobid = 10778349

| Epoch | c_loss | gp_loss | d_loss | cg_loss | recon_loss | g_loss | val_cg_loss | val_recon_loss |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Epoch 1 |  -3.0717 |  9.4734 |  6.4016 |  46.3662 |  422.8826 |  469.2488 |  52.3143 |  598.0638 |
| Epoch 2 |  -2.3982 |  9.0319 |  6.6337 |  93.9494 |  382.5220 |  476.4714 |  -15.6816 |  356.3661 |
| Epoch 3 |  -24.4850 |  5.6571 |  -18.8279 |  -130.4163 |  399.5670 |  269.1507 |  -80.8641 |  1393.8457 |
| Epoch 4 |  -134.6226 |  1.1733 |  -133.4493 |  -1160.3834 |  476.1442 |  -684.2392 |  -725.4740 |  9590.5293 |

**Comment by mlangguth89** [comment]
@seblehner 
Have you performed the tests with `verbose=1` to track the losses during the training iteration? If yes, can you please attach the corresponding files?

**Comment by seblehner** [comment]
@mlangguth89 Yes, here are the two corresponding files. 
[train_harris_wgan_ws100m_benchmark_test.sh-out.10778296.txt](https://github.com/user-attachments/files/18397952/train_harris_wgan_ws100m_benchmark_test.sh-out.10778296.txt)
[train_harris_wgan_ws100m_benchmark_test.sh-out.10778349.txt](https://github.com/user-attachments/files/18397953/train_harris_wgan_ws100m_benchmark_test.sh-out.10778349.txt)

**Comment by mlangguth89** [comment]
@seblehner 

As already shared today via Slack, there does not seem to be an issue in the data itself. 
Training a model with the config of Experiment 1 gives an RMSE of about 1.7 m/s for the ensemble mean (about 2.0 m/s for individual ensemble members). This is not very skillful as can be seen from comparison with #23, but it is in a reasonable value range.  Some example results are depicted below:

![Image](https://github.com/user-attachments/assets/12950a8a-89d7-4f7e-ba69-afc00973010e)
**Fig. 1:** Diurnal cycle of the domain-averaged RMSE averaged over test year 2018.

![Image](https://github.com/user-attachments/assets/e6449c19-6c26-49b4-ad11-aca58f1be938)
**Fig. 2:** Representative sample from test dataset showing the ground truth data (left), the downscaled product from an ensemble member of the Harris WGAN and the difference between both.

However a more severe general issue with the data normalization has been noted during evaluation. The outputted data spectrum is completely unrealistic. While (slightly) negative wind speed values may be acceptable since there is no corresponding constraint in the training and normalization process, the fairly common occurrence of excessive wind speed data is troublesome (see Fig. 3). 

![Image](https://github.com/user-attachments/assets/d303fe47-493a-479b-a845-caa140cd7614)
**Fig 3.:** Histogram of the ground truth wind speed data and the Harris WGAN downscaling product (ensemble member 1).

The reason for this behavior can be traced back to the PDF of the ground truth wind speed data which resembles a power law distribution (a bit similar to precipitation). However, the data is z-score normalized for training which is likely improper in this case. Thus, it might be advantageous to perform a log transformation to the windspeed data. In case of the Harris WGAN model, this would also allow to turn back to the softplus activation function used for the output layer which has been replaced by a linear output layer in #11.

**Comment by mlangguth89** [comment]
It looks like that there are spurious issues along the lateral boundaries which happens for some samples only. In Figure 1, the area-averaged RMSE is plotted over the test dataset (x-axis is the timestep axis). There are several samples with excessively high RMSE values, e.g. time step 89. 
![Image](https://github.com/user-attachments/assets/100f6d07-297d-410a-908f-0d318ac8eee9)
**Fig. 1**: Time-series of the area-averaged RMSE with time steps along the x-axis.

Closer inspection of these samples reveals that the Harris WGAN outputs small areas with very high wind speeds near the lateral boundaries of the domain. 

![Image](https://github.com/user-attachments/assets/6edec874-f918-4a6b-84ed-2c59e9db79f7)
**Fig. 2:** Forecast of ensemble member zero for 2018-01-01 18:00 time step 233 of the test dataset. Corresponding domain-averaged RMSE is 19.31 K.

The high error values furthermore only happen over certain areas as can be seen from the spatial RMSE-maps (see Fig. 3):

![Image](https://github.com/user-attachments/assets/349a486d-8641-49b9-87db-758b5ff8aa7a)

It is hypothesized that this also happens during training which in turn explains the high reconstruction loss values. 
The reasons for this are further investigated. 

Note: Evaluation results shown here are obtained from a run after 11 epochs of training.

**Comment by mlangguth89** [comment]
**NOTE**: Original plot of upscaled ERa5 data was wrong and has been corrected. 

Another update:

Training on log-transformed data is more stabl and reduces significantly the number of excessive windspeeds in the output data. However, training still becomes unstable (large negative critics after ~5 epochs) and reconstruction loss does not go down after first epoch. 

Further investigation on the input data now reveals issues as illustrated below from a sample of the test dataset:
For 2018-01 16:00 UTC, the COSMO REA 6 windspeed data is as follows:

![Image](https://github.com/user-attachments/assets/58f77197-3418-4a6e-aa4e-b048b705cc64)

The ERA5 input 100m windspeed data look quite different. First, the coarse-grained input data as used by the Harris WGAN is displayed:

![Image](https://github.com/user-attachments/assets/b82dab64-316a-45b1-854c-c7577a26c501)

Second, the bilinearly updscaled ERA5 input 100m windspeed data (as used by the Sha WGAN, SwinIR and Sha U-Net) is shown:

![Image](https://github.com/user-attachments/assets/baa34abe-7628-45f9-b0d1-1c1ffb9a0090)

The bilinearly upscaled and the coarse ERA5 data are consistent.
However. especially along the lateral boundaries, but also in the inner parts, substantial differences are present.
Double checking with manual data process is required.

**Comment by mlangguth89** [comment]
~~Seems to be an error related to using CDO in an non-thread safe environment. ~~
~~I've just created the following sliced ground truth data from COSMO REA6 which resembles strongly the bi-linearly upscaled ERA5 data shown above:~~

Manual re-processing of the ERA5-data confirms correctness of the data (at least for this sample on 2018-01-03 15:00 UTC).
An time index-shift must also be considered when processing ERA% data only since the dataset starts at 00 UTC (CREA6 starts at 01 UTC).

![Image](https://github.com/user-attachments/assets/baf8ac2f-a15b-4c8b-8ba3-24fd0bafb2c2)

**Comment by mlangguth89** [comment]
Looks like that the whole preprocessing must be re-run in a thread-safe environment, which I will try to realizse shortly.

**Comment by mlangguth89** [comment]
Independant checking against plots of ERA5 and COSMO REA6 also rather confirm the correctness of the data:

10m wind from ERA5: https://kachelmannwetter.com/de/reanalyse/ecmwf-era5/bcf1e1649f215a93c7eb65f382c4a689/windgeschwindigkeit-richtung-ms/20180103-1500z.html

10m wind from COSMO REA6: https://kachelmannwetter.com/de/reanalyse/cosmo-rea6/bcf1e1649f215a93c7eb65f382c4a689/windrichtung-windmittel/20180103-1500z.html

In this case, there is obviously a systematic difference between both reanalysis products over the Alpine region (with stronger winds in CREA6).

Further experiments with the log-transformed data, e.g. varying the weight of the reconstruction loss are on-going. Those are documented under [this link](https://gitlab.jsc.fz-juelich.de/hedgedoc/syjpz3eBTq2ckXgVpoTDzg?view#).

**Comment by seblehner** [comment]
Just for documentation, here some independet checks about the wind data.

- took the already preprocessed training data and the original era5 wind data; specifically `ws100_squared` for the original data and `ws100m_in` from the training data
- regridded the original data via cdo: `cdo -L remapcon,/p/scratch/hclimrep/maelstrom/downscaling/downscaling_benchmark_dataset/benchmark_wind/preprocessed/coarse_input/crea6_coarsened_grid ws100_sq_orig.nc ws100_sq_orig_reproj.nc` and discarded 1 gridpoint boundary around the domain (such that it's similar to the already processed training data)
- then simply calculated the squared error between the original, regridded `ws100_squared` and the `ws100m_in` squared fields for all hourly timesteps and the full available congruent time period
- the errors are then spatially reduced by a) mean and b) max, yielding two time series which are shown below
- as can be seen in the plot, the errors are incredibly small (notice the large negative exponents on the y-scale) and hence negligible, which means the manually processed data is similar to the one from the processing pipeline

![Image](https://github.com/user-attachments/assets/2e7210dd-2e90-4960-a6dc-b187b7cf8d7b)
